### PR TITLE
all: Lint Python code with ruff.

### PR DESCRIPTION
**Pull request by cclauss**
As hinted at https://github.com/micropython/micropython/pull/10977#issuecomment-1531559631

[Ruff](https://beta.ruff.rs/) supports [over 500 lint rules](https://beta.ruff.rs/docs/rules) and can be used to replace [Flake8](https://pypi.org/project/flake8/) (plus dozens of plugins), [isort](https://pypi.org/project/isort/), [pydocstyle](https://pypi.org/project/pydocstyle/), [yesqa](https://github.com/asottile/yesqa), [eradicate](https://pypi.org/project/eradicate/), [pyupgrade](https://pypi.org/project/pyupgrade/), and [autoflake](https://pypi.org/project/autoflake/), all while executing (in Rust) tens or hundreds of times faster than any individual tool.

The ruff Action uses minimal steps to run in ~5 seconds, rapidly providing intuitive GitHub Annotations to contributors.

image

### CODE REVIEW

File: micropython/mip/mip/__init__.py [hunk 1]
```diff
@@ -123,12 +123,8 @@ def _install_json(package_json_url, index, target, version, mpy):
...
 
 def _install_package(package, index, target, version, mpy):
-    if (
-        package.startswith("http://")
-        or package.startswith("https://")
-        or package.startswith("github:")
-    ):
-        if package.endswith(".py") or package.endswith(".mpy"):
+    if package.startswith(("http://", "https://", "github:")):
+        if package.endswith((".py", ".mpy")):
```
Comment (dpgeorge) [inline micropython/mip/mip/__init__.py:127]: MicroPython doesn't support tuples of strings passed to startswith/endswith, so please revert this change.

Can ruff be configured to exclude this particular warning?

**Comment by dpgeorge** [comment]
Thanks for working on this!

File: .github/workflows/build_packages.yml [hunk 2]
```diff
@@ -9,8 +9,9 @@ jobs:
   build:
     runs-on: ubuntu-latest
     steps:
-    - uses: actions/checkout@v2
-    - uses: actions/setup-python@v1
+    - uses: actions/checkout@v3
+    - uses: actions/setup-python@v4
+      with: {python-version: 3.x}
```
[the remark this answers is not in the document]
Comment (cclauss) [reply .github/workflows/build_packages.yml:14]: It is in all the examples https://github.com/actions/setup-python/blob/main/docs/advanced-usage.md

**Comment by cclauss** [comment]
Gentle bump...

**Comment by cclauss** [comment]
Gentle bump...

**Review by dpgeorge — APPROVED** [review]

**Comment by dpgeorge** [comment]
Thanks for the reminder.

I rebased and squashed the changes.  But looks like ruff was updated in the meantime and now has more warnings?

Note that some of the suggestions go against MicroPython best practice, eg because they allocate memory.

**Comment by cclauss** [comment]
OK...  Rebased.  I do not see any memory allocations but please let me know if you do.

File: micropython/lora/examples/reliable_delivery/sender.py [hunk 3]
```diff
@@ -149,7 +149,7 @@ def send(self, sensor_data, adjust_output_power=True):
                 delta = time.ticks_diff(maybe_ack.ticks_ms, sent_at)
                 print(
                     f"ACKed with RSSI {rssi}, {delta}ms after sent "
-                    + f"(skew {delta-ACK_DELAY_MS-ack_packet_ms}ms)"
+                    f"(skew {delta-ACK_DELAY_MS-ack_packet_ms}ms)"
```
Comment (dpgeorge) [inline micropython/lora/examples/reliable_delivery/sender.py:152]: MicroPython doesn't support adjacent f-strings, it needs the "+".

(This file is not tested by the CI here, because it's an example.)

File: micropython/lora/examples/reliable_delivery/sender_async.py [hunk 4]
```diff
@@ -141,7 +141,7 @@ async def send(self, sensor_data, adjust_output_power=True):
                 delta = time.ticks_diff(maybe_ack.ticks_ms, sent_at)
                 print(
                     f"ACKed with RSSI {rssi}, {delta}ms after sent "
-                    + f"(skew {delta-ACK_DELAY_MS-ack_packet_ms}ms)"
+                    f"(skew {delta-ACK_DELAY_MS-ack_packet_ms}ms)"
```
Comment (dpgeorge) [inline micropython/lora/examples/reliable_delivery/sender_async.py:144]: The "+" is needed.

File: micropython/lora/tests/test_time_on_air.py [hunk 5]
```diff
@@ -183,7 +183,7 @@ def main():
         (sf, bw_khz, coding_rate, low_datarate_optimize) = modem_settings
         print(
             f"Modem config sf={sf} bw={bw_khz}kHz coding_rate=4/{coding_rate} "
-            + f"low_datarate_optimize={low_datarate_optimize}"
+            f"low_datarate_optimize={low_datarate_optimize}"
```
Comment (dpgeorge) [inline micropython/lora/tests/test_time_on_air.py:186]: The "+" is needed.

**Comment by cclauss** [comment]
Ignore `ruff rule ISC003`  # micropython does not support implicit concatenation of f-strings
# explicit-string-concatenation (ISC003)

Derived from the **flake8-implicit-str-concat** linter.

## What it does
Checks for string literals that are explicitly concatenated (using the
`+` operator).

## Why is this bad?
For string literals that wrap across multiple lines, implicit string
concatenation within parentheses is preferred over explicit
concatenation using the `+` operator, as the former is more readable.

## Example
```python
z = (
    "The quick brown fox jumps over the lazy "
    + "dog"
)
```

Use instead:
```python
z = (
    "The quick brown fox jumps over the lazy "
    "dog"
)
```

**Comment by dpgeorge** [comment]
Thank you, looks pretty good now!
